src/TriggerBinding/SqlTriggerListener.cs

```diff
@@ -0,0 +1,356 @@
+            _ = !string.IsNullOrEmpty(userFunctionId) ? true : throw new ArgumentNullException(nameof(userFunctionId));
+            _ = executor ?? throw new ArgumentNullException(nameof(executor));
+            _ = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            this._connectionString = connectionString;
+            this._userTable = new SqlObject(tableName);
+            this._userFunctionId = userFunctionId;
+            this._executor = executor;
+            this._logger = logger;
+            this._listenerState = ListenerNotStarted;
+        }
+
+        public void Cancel()
+        {
+            this.StopAsync(CancellationToken.None).GetAwaiter().GetResult();
+        }
+
+        public void Dispose()
+        {
+            // Nothing to dispose.
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            int previousState = Interlocked.CompareExchange(ref this._listenerState, ListenerStarting, ListenerNotStarted);
+
+            switch (previousState)
+            {
+                case ListenerStarting: throw new InvalidOperationException("The listener is already starting.");
+                case ListenerStarted: throw new InvalidOperationException("The listener has already started.");
+                default: break;
+            }
+
+            try
+            {
+                using (var connection = new SqlConnection(this._connectionString))
+                {
+                    await connection.OpenAsync(cancellationToken);
+
+                    int userTableId = await this.GetUserTableIdAsync(connection, cancellationToken);
+                    IReadOnlyList<(string name, string type)> primaryKeyColumns = await this.GetPrimaryKeyColumnsAsync(connection, userTableId, cancellationToken);
+                    IReadOnlyList<string> userTableColumns = await GetUserTableColumnsAsync(connection, userTableId, cancellationToken);
+
+                    string workerTableName = string.Format(CultureInfo.InvariantCulture, SqlTriggerConstants.WorkerTableNameFormat, $"{this._userFunctionId}_{userTableId}");
+
+                    using (SqlTransaction transaction = connection.BeginTransaction(System.Data.IsolationLevel.RepeatableRead))
+                    {
+                        await CreateSchemaAsync(connection, transaction, cancellationToken);
+                        await CreateGlobalStateTableAsync(connection, transaction, cancellationToken);
+                        await this.InsertGlobalStateTableRowAsync(connection, transaction, userTableId, cancellationToken);
+                        await CreateWorkerTablesAsync(connection, transaction, workerTableName, primaryKeyColumns, cancellationToken);
+                        transaction.Commit();
+                    }
+
+                    // TODO: Check if passing the cancellation token would be beneficial.
```

Comment:
This seems like something we should be doing

Reply:
Leaving the TODO comment for now and adding the task to #282.